cranelift-codegen/src/isa/x86/binemit.rs

```diff
@@ -101,63 +103,61 @@ fn put_rexop2<CS: CodeSink + ?Sized>(bits: u16, rex: u8, sink: &mut CS) {
 
 // Emit single-byte opcode with mandatory prefix.
 fn put_mp1<CS: CodeSink + ?Sized>(bits: u16, rex: u8, sink: &mut CS) {
+    let enc = EncodingBits::from(bits);
     debug_assert_eq!(bits & 0x8c00, 0, "Invalid encoding bits for Mp1*");
-    let pp = (bits >> 8) & 3;
-    sink.put1(PREFIX[(pp - 1) as usize]);
+    sink.put1(PREFIX[(enc.pp() - 1) as usize]);
     debug_assert_eq!(rex, BASE_REX, "Invalid registers for REX-less Mp1 encoding");
     sink.put1(bits as u8);
 }
 
 // Emit single-byte opcode with mandatory prefix and REX.
 fn put_rexmp1<CS: CodeSink + ?Sized>(bits: u16, rex: u8, sink: &mut CS) {
+    let enc = EncodingBits::from(bits);
     debug_assert_eq!(bits & 0x0c00, 0, "Invalid encoding bits for Mp1*");
-    let pp = (bits >> 8) & 3;
-    sink.put1(PREFIX[(pp - 1) as usize]);
+    sink.put1(PREFIX[(enc.pp() - 1) as usize]);
```

Comment:
Seems like the `let enc = ...` should come after the check to see if it the bits are valid; I think even cleaner would be to somehow use the new `OpcodePrefix` code to do the check instead of a bitwise `&`.

Reply:
Unfortunately it's not just checking the OpcodePrefix (which is `0xf00`) -- it also checks the `rrr` bits, as in `0x8c00` for `Mp2`. It's hard to give a name to that. We'll be rewriting all these methods eventually as part of recipe refactoring, so it should be OK to put that off.